### PR TITLE
Allow customers to override Journeys CTA links - JOUR-3883

### PR DESCRIPTION
**Pull request by jwangler-branch**
- setBranchViewData() should allow the user to specify a $journeys_cta parameter. Witch overrides the branch link entirely

- If the user tries to update link redirects using existing Branch params (e.g. $android_url, $ios_url, $fallback_url) then we should allow that.

### CODE REVIEW

File: src/journeys_utils.js [hunk 1]
```diff
@@ -908,3 +908,72 @@ journeys_utils.setJourneyLinkData = function(linkData) {
 	}
 	journeys_utils.journeyLinkData = data;
 };
+
+journeys_utils.hasInBranchViewData = function(fieldName) {
```
Comment (hemanth-s-a) [inline src/journeys_utils.js]: What does the name of the method mean?

Comment (jwangler-branch) [reply src/journeys_utils.js]: Ok, bad naming. :) The function returns **true** if in BranchViewData a field with the name: FieldName

Comment (jwangler-branch) [reply src/journeys_utils.js]: a better name could be `hasBranchViewData(fieldName) or isFieldInBranchViewData(fieldName) or  doesParameterExitsInBranchViewData(parameterName) or exitsBranchViewData(name)`

Comment (jwangler-branch) [reply src/journeys_utils.js]: `exitsValueInBranchViewData(name)`

**Comment by rubinsingh** [comment]
@jwangler-branch could you please add build.js and build.min.js to the gitignore file? I'd also remove them from your PR.

File: dist/build.js [hunk 2]
```diff
@@ -2799,11 +2843,15 @@ Branch.prototype.logEvent = wrap(callback_params.CALLBACK_ERR, function(a, b, c,
   });
 });
 Branch.prototype.link = wrap(callback_params.CALLBACK_ERR_DATA, function(a, b) {
-  var c = utils.cleanLinkData(b), d = this.branch_key;
+  var c = utils.cleanLinkData(b), c = journeys_utils.trySetBranchViewDataUrls(b), d = this.branch_key;
   this._api(resources.link, c, function(b, f) {
     if (b) {
       return a(b, utils.generateDynamicBNCLink(d, c));
     }
+    console.log(33);
```
Comment (rubinsingh) [inline dist/build.js]: console.log()

File: src/journeys_utils.js [hunk 3]
```diff
@@ -908,3 +908,72 @@ journeys_utils.setJourneyLinkData = function(linkData) {
...
+			}
+			return value;
+		}, data);
+	};
+
+	try {
+		var data = (JSON.parse(linkElements.data));
+		linkElements.data = JSON.stringify(assignUrls(data));
+	}catch(e){
+		console.log(e);
```
Comment (rubinsingh) [inline src/journeys_utils.js]: Are we sure we want to print errors to the console?

File: src/journeys_utils.js [hunk 4]
```diff
@@ -908,3 +908,72 @@ journeys_utils.setJourneyLinkData = function(linkData) {
...
+	 return undefined;
+};
+
+journeys_utils.getJourneyCtaLink = function () {
+	return journeys_utils.getBranchViewDataItemOrUndefined('$journeys_cta');
+};
+
+journeys_utils.trySetBranchViewDataUrls = function (linkElements, urls = ['$android_url', '$ios_url', '$fallback_url', '$desktop_url']) {
+	if(!linkElements){
+		console.log("setDefaultUrls - no data set");
```
Comment (rubinsingh) [inline src/journeys_utils.js]: console.log()

File: src/journeys_utils.js [hunk 5]
```diff
@@ -908,3 +908,72 @@ journeys_utils.setJourneyLinkData = function(linkData) {
...
+		return linkElements;
+	}
+
+	var assignUrls = function (data) {
+		return urls.reduce((value, url)=>{
```
Comment (rubinsingh) [inline src/journeys_utils.js]: Does ES6 code get minified properly?

File: src/6_branch.js [hunk 6]
```diff
@@ -1252,12 +1252,19 @@ Branch.prototype['logEvent'] = wrap(callback_params.CALLBACK_ERR, function(done,
 /*** +TOC_ITEM #linkdata-callback &.link()& ^ALL ***/
 Branch.prototype['link'] = wrap(callback_params.CALLBACK_ERR_DATA, function(done, data) {
 	var linkData = utils.cleanLinkData(data);
+	linkData = journeys_utils.trySetBranchViewDataUrls(data);
```
Comment (rubinsingh) [inline src/6_branch.js]: Can we change the name of this function? I'm also confused as to why we're modifying the .link() function if the code is supposed to affect the Journeys CTA only?

File: src/journeys_utils.js [hunk 7]
```diff
@@ -908,3 +908,72 @@ journeys_utils.setJourneyLinkData = function(linkData) {
 	}
 	journeys_utils.journeyLinkData = data;
 };
+
+journeys_utils.exitsValueInBranchViewData = function(name) {
```
Comment (rubinsingh) [inline src/journeys_utils.js]: exitsValueInBranchViewData -> getValueForKeyInBranchViewData = function(key)

Comment (jwangler-branch) [reply src/journeys_utils.js]: 👍

File: src/6_branch.js (hunk 6 above)
Comment (jwangler-branch) [reply src/6_branch.js]: 👍

File: src/journeys_utils.js (hunk 4 above)
Comment (jwangler-branch) [reply src/journeys_utils.js]: 👍

File: src/journeys_utils.js (hunk 3 above)
Comment (jwangler-branch) [reply src/journeys_utils.js]: In some way, we have to notify someone if there is an exception. We don't have a logging system for the SDK so only console log works. But I have an idea to deal with that. I'll try something 👍

File: dist/build.js (hunk 2 above)
Comment (jwangler-branch) [reply dist/build.js]: 👍

File: src/journeys_utils.js (hunk 5 above)
Comment (jwangler-branch) [reply src/journeys_utils.js]: Yes it works properly

**Pull request closed without merging**